File: haskell/P0001/sum-multiples/src/Solution.hs
module Solution
    ( sumAllMultiplesOfBelow
    ) where

isMultipleOf :: Int -> Int -> Bool
isMultipleOf n m = n `mod` m == 0

sumAllMultiplesOfBelow :: Int -> Int -> Int -> Int
sumAllMultiplesOfBelow a b below = sum(filter (\n -> isMultipleOf n a || isMultipleOf n b) [1..below - 1])

File: rust/P0001/sum-multiples/src/solution.rs
pub fn sum_all_multiples_of_below(a: u128, b: u128, below: u128) -> u128 {
    return (1..below)
      .filter(|n| n % a == 0 || n % b ==0)
      .fold(0, |acc, i| acc + i); 
}

#[cfg(test)]
mod tests {
    // Note this useful idiom: importing names from outer (for mod tests) scope.
    use super::*;

    #[test]
    fn test_sum_all_multiples_of_3_and_5_below_10() {
        assert_eq!(sum_all_multiples_of_below(3, 5, 10), 23);
    }

    #[test]
    fn test_sum_all_multiples_of_3_and_5_below_1000() {
        assert_eq!(sum_all_multiples_of_below(3, 5, 1000), 233168);
    }
}